examples/chat2-reliable/reliability.go

```diff
@@ -0,0 +1,458 @@
+	bloomFilterCleanInterval = 30 * time.Minute
+	bufferSweepInterval      = 5 * time.Second
+	syncMessageInterval      = 60 * time.Second
+	messageAckTimeout        = 10 * time.Second
+	maxRetries               = 3
+	retryBaseDelay           = 1 * time.Second
+	maxRetryDelay            = 10 * time.Second
+	ackTimeout               = 5 * time.Second
+	maxResendAttempts        = 5
+	resendBaseDelay          = 1 * time.Second
+	maxResendDelay           = 30 * time.Second
+)
+
+func (c *Chat) initReliabilityProtocol() {
+	c.wg.Add(5)
+	go c.periodicBufferSweep()
+	go c.periodicSyncMessage()
+	go c.setupMessageRequestHandler()
+	go c.startBloomFilterCleaner()
+	go c.startEagerPushMechanism()
+}
+
+func (c *Chat) startEagerPushMechanism() {
+	defer c.wg.Done()
+
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		case <-ticker.C:
+			c.checkUnacknowledgedMessages()
+		}
+	}
+}
+
+type UnacknowledgedMessage struct {
+	Message        *pb.Message
+	SendTime       time.Time
+	ResendAttempts int
+}
+
+type TimestampedMessageID struct {
+	ID        string
+	Timestamp time.Time
+}
+
+type RollingBloomFilter struct {
+	filter   *bloom.BloomFilter
+	window   time.Duration
+	messages []TimestampedMessageID
+	mutex    sync.Mutex
+}
+
+func NewRollingBloomFilter() *RollingBloomFilter {
+	return &RollingBloomFilter{
+		filter:   bloom.NewWithEstimates(bloomFilterSize, bloomFilterFPRate),
+		window:   bloomFilterWindow,
+		messages: make([]TimestampedMessageID, 0),
+	}
+}
+
+func (rbf *RollingBloomFilter) Add(messageID string) {
+	rbf.mutex.Lock()
+	defer rbf.mutex.Unlock()
+
+	rbf.filter.Add([]byte(messageID))
+	rbf.messages = append(rbf.messages, TimestampedMessageID{
+		ID:        messageID,
+		Timestamp: time.Now(),
+	})
+}
+
+func (rbf *RollingBloomFilter) Test(messageID string) bool {
+	rbf.mutex.Lock()
+	defer rbf.mutex.Unlock()
+
+	return rbf.filter.Test([]byte(messageID))
+}
+
+func (rbf *RollingBloomFilter) Clean() {
+	rbf.mutex.Lock()
+	defer rbf.mutex.Unlock()
+
+	cutoff := time.Now().Add(-rbf.window)
+	newMessages := make([]TimestampedMessageID, 0)
+	newFilter := bloom.NewWithEstimates(bloomFilterSize, bloomFilterFPRate)
+
+	for _, msg := range rbf.messages {
+		if msg.Timestamp.After(cutoff) {
+			newMessages = append(newMessages, msg)
+			newFilter.Add([]byte(msg.ID))
+		}
+	}
+
+	rbf.messages = newMessages
+	rbf.filter = newFilter
+}
+
+func (c *Chat) startBloomFilterCleaner() {
+	defer c.wg.Done()
+
+	ticker := time.NewTicker(bloomFilterCleanInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		case <-ticker.C:
+			c.bloomFilter.Clean()
+		}
+	}
+}
+
+func (c *Chat) processReceivedMessage(msg *pb.Message) {
+	// Check if the message is already in the bloom filter
+	if c.bloomFilter.Test(msg.MessageId) {
+		// Review ACK status of messages in the unacknowledged outgoing buffer
+		c.reviewAckStatus(msg)
+		return
+	}
```

Comment:
It feels like I might be missing something obvious, but why only review ACK status of outgoing messages if the received ID is already in the bloom filter? Shouldn't it be the opposite, i.e. for every newly received message we want to check if it happens to ACK some of our outgoing messages?